kernel/src/hil/time.rs

```diff
@@ -1,33 +1,276 @@
+/// counter has reached a pre-specified value set in
+/// [`set_alarm`](#tymethod.set_alarm). Alarms are intended for
+/// low-level time needs that require precision (i.e., firing on a
+/// precise clock tick). Software that needs more functionality
+/// but can tolerate some jitter should use the `Timer` trait
+/// instead.
+pub trait Alarm<'a>: Time {
+    /// Specify the callback for when the counter reaches the alarm
+    /// value. If there was a previously installed callback this call
+    /// replaces it.
+    fn set_alarm_client(&'a self, client: &'a dyn AlarmClient);
+
+    /// Specify when the callback should be called and enable it. The
+    /// callback will be enqueued when `Time::now() == reference + dt`. The
+    /// callback itself may not run exactly at this time, due to delays.
+    /// However, it it assured to execute *after* `reference + dt`: it can
+    /// be delayed but will never fire early. The method takes `reference`
+    /// and `dt` rather than a single value denoting the counter value so it
+    /// can distinguish between alarms which have very recently already
+    /// passed and those in the far far future (see #1651).
+    fn set_alarm(&self, reference: Self::Ticks, dt: Self::Ticks);
+
+    /// Return the current alarm value. This is undefined at boot and
+    /// otherwise returns `now + dt` from the last call to `set_alarm`.
+    fn get_alarm(&self) -> Self::Ticks;
+
+    /// Disable the alarm and stop it from firing in the future.
+    /// Valid `ReturnCode` codes are:
+    ///   - `ReturnCode::SUCCESS` the alarm has been disarmed and will not invoke
+    ///   the callback in the future    
+    ///   - `ReturnCode::FAIL` the alarm could not be disarmed and will invoke
+    ///   the callback in the future    
+    fn disarm(&self) -> ReturnCode;
+
+    /// Returns whether the alarm is currently armed. Note that this
+    /// does not reliably indicate whether there will be a future
+    /// callback: it is possible that the alarm has triggered (and
+    /// disarmed) and a callback is pending and has not been called yet.
+    /// In this case it possible for `is_armed` to return false yet to
+    /// receive a callback.
+    fn is_armed(&self) -> bool;
+
+    /// Return the minimum dt value that is supported. Any dt smaller than
+    /// this will automatically be increased to this minimum value.
+    fn minimum_dt(&self) -> Self::Ticks;
+}
+
+/// Callback handler for when a timer fires.
+pub trait TimerClient {
+    fn timer(&self);
+}
+
+/// Interface for controlling callbacks when an interval has passed.
+/// This interface is intended for software that requires repeated
+/// and/or one-shot timers and is willing to experience some jitter or
+/// imprecision in return for a simpler API that doesn't require
+/// actual calculation of counter values. Software that requires more
+/// precisely timed callbacks should use the `Alarm` trait instead.
+pub trait Timer<'a>: Time {
```

Comment:
I don't think that this supertrait bound to `Time` is justified. A `Timer` is a device operating on a relative time interval, namely the one from enabling until the specified number of ticks have been counted down. Thereby it does not necessarily have a notion of "absolute" (monotonically increasing) time as mandated by the `Time` trait. An example for hardware that implements solely a `Timer` without an "uptime value" is the LiteX Timer core in its default configuration.

I have [implemented a proposed fix here](https://github.com/lschuermann/tock/commit/3f9892c336efc9a5c650594e546bf17309adbdce), feel free to pull this in or request changes.

This actually solves a second problem with my hardware: **if** (not always the case) configured with a system clock, its width is `u64` whereas the timer will be `u32` by default. Therefore it is important to have `Ticks` (and possibly `Frequency`) independent from the `Time` trait as well.

For `Alarm` both of the constraints make sense. Setting an alarm to a timestamp in the future necessarily requires knowing the current time and the alarm time must be in the same field (word size) as the time.

Reply:
Not changing the APIs at the zeroth hour of final testing. 

It inherits from Time in part because this gives it Ticks and Frequency. 

Why is its width u32 by default?

Reply:
> Not changing the APIs at the zeroth hour of final testing.

Makes sense. I am sorry that this came up so late and I was not aware of the plans on merging in the (very) near future. I can open another PR once this one it merged. My "fix" does not necessarily need to meet any release deadlines and therefore has a bit of time.

> It inherits from Time in part because this gives it Ticks and Frequency.

I see. If that is the _only_ reason, my suspicion of `Timer` not requiring a measure of "absolute" time, and hence my fix does seem to be correct.

> Why is its width u32 by default?

LiteX Timers (`Timer`-trait) simply have a default width of 32bits. Enabling the uptime feature on LiteX SoCs adds a 64bit uptime (`Time`-trait) counter to this timer. I'd like to allow Tock to run on the defaults.

I don't however think that this is the issue here, just an instance of where the flexibility of the current HIL is insufficient. I don't have other specific examples but could think that other chips might have different word-widths for different time/alarm/timer peripherals attached to different clocks or even an integrated RTC. I think we should allow this flexibility which is as simple as adding a `Ticks` type to the timer independent of `Time`/`Alarm` to allow independent implementation of a `Timer` without a `Time` (whereas `Ticks` and `Frequency` of `Time` and `Alarm` should always coincide since an `Alarm` must always have a notion of time).

Reply:
I can't seem to take this "change suggestion" back so ignore please. :)

Reply:
A `Timer` can need `Time` beyond `Frequency` and `Ticks`. They aren't necessarily always operating in terms of a relative time. For example, it could be that a network protocol does an exchange to decide on a regular communication interval. It needs to be able to set not only the frequency (the interval) but also its phase. By removing `Time`, you're unnecessarily constraining `Timer` only to settings in which you do not require any notion of a shared or global time base. 

Another example use case: your software wants to sleep/do nothing until a particular point in future time (sync to a phase) then start a periodic timer that fires 5 times. E.g., wants to wait until 10:23 then sample temperature 5 times at 1Hz.

Reply:
I get your point. I don't however think that the current `Timer` trait provides us sufficient methods to guarantee a constant phase offset with respect to a global/shared time, since we can only set it in terms of relative (countdown) time. I can imagine a Timer with constant phase offset with respect to a global/shared time can be much better built using a repeatedly set Alarm, since I can set the timer to an absolute value of the global/shared time. Currently `Timer` only offers a constant phase offset with respect to the previous callback, namely using the `repeating()` method and given HW support.

More pressing for me is what I should do in my scenario. I do have a `Timer`-compatible HW peripheral, but that is not also a `Time`-compatible peripheral. Which combination of traits could abstract this peripheral?

What would you think of a trait `FixedPhaseTimer: Time + Timer` that could be implemented given sufficient HW support and actually provides a method such as `phase_repeating(start: Self::Ticks, interval: Self::Ticks)`? I can imagine that being very useful in my upcoming work.

Reply:
I guess I don't understand how your hardware is incompatible. It's OK to have multiple implementations of Time/Counter. E.g., you could have one module that reads it as 64 bits, another that reads it as 32. Can you point me at the sections of the datasheet so I can read and understand better?

Reply:
> Can you point me at the sections of the datasheet so I can read and understand better?

Since LiteX is an SoC-Generator, it unfortunately does not provide a "datasheet" in the classical sense.

My intention was not to hijack this PR. Since any potential change won't be included here, I will open an issue later this week where we can discuss whether the current traits can be used for my hardware. I will attach a table of timer registers and explain how the peripheral works.

Reply:
@phil-levis After close inspection of the new HIL, I think that my issues can be worked around. In particular while the traits in question have not changed much, the fact that a `Time` implementation is not semantically equivalent to a `Timestamp` anymore resolves the primary issue of having to implement `Time` for `Timer`. Please excuse my confusion which mostly originated from the legacy (and thankfully) replaced HIL.